mithril-stm/src/merkle_tree/tree.rs

```diff
@@ -492,127 +184,56 @@ impl<D: Digest + FixedOutput> MerkleTree<D> {
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct MerkleTreeCommitment<D: Digest> {
-    /// Root of the merkle commitment.
-    pub root: Vec<u8>,
-    hasher: PhantomData<D>,
-}
+    /// Get a path (hashes of siblings of the path to the root node)
+    /// for the `i`th value stored in the tree.
+    /// Requires `i < self.n`
+    pub fn get_path(&self, i: usize) -> Path<D> {
+        assert!(
+            i < self.n,
+            "Proof index out of bounds: asked for {} out of {}",
+            i,
+            self.n
+        );
+        let mut idx = self.idx_of_leaf(i);
+        let mut proof = Vec::new();
 
-impl<D: Clone + Digest + FixedOutput> MerkleTreeCommitment<D> {
-    /// Check an inclusion proof that `val` is part of the tree by traveling the whole path until the root.
-    /// # Error
-    /// If the merkle tree path is invalid, then the function fails.
-    pub fn check(&self, val: &MTLeaf, proof: &Path<D>) -> Result<(), MerkleTreeError<D>> {
-        let mut idx = proof.index;
-
-        let mut h = D::digest(val.to_bytes()).to_vec();
-        for p in &proof.values {
-            if (idx & 0b1) == 0 {
-                h = D::new().chain_update(h).chain_update(p).finalize().to_vec();
+        while idx > 0 {
+            let h = if sibling(idx) < self.nodes.len() {
+                self.nodes[sibling(idx)].clone()
             } else {
-                h = D::new().chain_update(p).chain_update(h).finalize().to_vec();
-            }
-            idx >>= 1;
-        }
-
-        if h == self.root {
-            return Ok(());
+                D::digest([0u8]).to_vec()
+            };
+            proof.push(h.clone());
+            idx = parent(idx);
         }
-        Err(MerkleTreeError::PathInvalid(proof.clone()))
-    }
-
-    /// Serializes the Merkle Tree commitment together with a message in a single vector of bytes.
-    /// Outputs `msg || self` as a vector of bytes.
-    pub fn concat_with_msg(&self, msg: &[u8]) -> Vec<u8>
-    where
-        D: Digest,
-    {
-        let mut msgp = msg.to_vec();
-        let mut bytes = self.root.clone();
-        msgp.append(&mut bytes);
 
-        msgp
+        Path::new(proof, i)
     }
 }
 
 // ---------------------------------------------------------------------
 // Heap Helpers
 // ---------------------------------------------------------------------
-fn parent(i: usize) -> usize {
+pub(crate) fn parent(i: usize) -> usize {
     assert!(i > 0, "The root node does not have a parent");
     (i - 1) / 2
 }
 
-fn left_child(i: usize) -> usize {
+pub(crate) fn left_child(i: usize) -> usize {
     (2 * i) + 1
 }
 
-fn right_child(i: usize) -> usize {
+pub(crate) fn right_child(i: usize) -> usize {
     (2 * i) + 2
 }
 
-fn sibling(i: usize) -> usize {
+pub(crate) fn sibling(i: usize) -> usize {
     assert!(i > 0, "The root node does not have a sibling");
     // In the heap representation, the left sibling is always odd
     // And the right sibling is the next node
```

Comment:
Maybe these function should be moved to `mod.rs` or `utils.rs` module of `merkle_tree` (and visibility for the crate would be necessary)?